<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

# Lab 4.1.1 Linear Regression

Prediction of House Price Using Linear Regression

# Data

The Sacramento real estate transactions file is a list of 985 real estate transactions in the Sacramento area reported over a five-day period, as reported by the Sacramento Bee.

In [1]:
import pandas as pd
import numpy as np
import pandas as pd

%matplotlib inline

## Explore Data

Perform EDA on this dataset. Use what you have learned so far. Be creative. List all your findings.

In [2]:
# Read CSV
house_csv = "C:/Users/user/Documents/IOD_data_MAIN/DATA/Sacramento_transactions.csv"
house = pd.read_csv(house_csv)

In [3]:
# Data Types
# Hint: Look into the columns zip, sale_date & beds
house.dtypes

street        object
city          object
zip            int64
state         object
beds           int64
baths          int64
sq__ft         int64
type          object
sale_date     object
price          int64
latitude     float64
longitude    float64
dtype: object

In [4]:
house.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879  
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028  
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839  
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146  
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768

In [5]:
house.shape

(985, 12)

In [6]:
house.size

11820

In [7]:
house.describe()

zip        beds       baths       sq__ft          price  \
count    985.000000  985.000000  985.000000   985.000000     985.000000   
mean   95750.697462    2.911675    1.776650  1314.916751  234144.263959   
std       85.176072    1.307932    0.895371   853.048243  138365.839085   
min    95603.000000    0.000000    0.000000     0.000000    1551.000000   
25%    95660.000000    2.000000    1.000000   952.000000  145000.000000   
50%    95762.000000    3.000000    2.000000  1304.000000  213750.000000   
75%    95828.000000    4.000000    2.000000  1718.000000  300000.000000   
max    95864.000000    8.000000    5.000000  5822.000000  884790.000000   

         latitude   longitude  
count  985.000000  985.000000  
mean    38.607732 -121.355982  
std      0.145433    0.138278  
min     38.241514 -121.551704  
25%     38.482717 -121.446127  
50%     38.626582 -121.376220  
75%     38.695589 -121.295778  
max     39.020808 -120.597599

In [8]:
house.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 985 entries, 0 to 984
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   street     985 non-null    object 
 1   city       985 non-null    object 
 2   zip        985 non-null    int64  
 3   state      985 non-null    object 
 4   beds       985 non-null    int64  
 5   baths      985 non-null    int64  
 6   sq__ft     985 non-null    int64  
 7   type       985 non-null    object 
 8   sale_date  985 non-null    object 
 9   price      985 non-null    int64  
 10  latitude   985 non-null    float64
 11  longitude  985 non-null    float64
dtypes: float64(2), int64(5), object(5)
memory usage: 92.5+ KB


In [ ]:
# sale_date column is wrong, there might be duplicated zip number if u check index 2,3.
# maybe the beds number is wrong and doesn't tally with the price 

In [ ]:
# Feature Engineering


In [10]:
#house.sale_date = pd.to_datetime(house.sale_date) 

C:\ProgramData\Anaconda3\lib\site-packages\dateutil\parser\_parser.py:1213: UnknownTimezoneWarning: tzname EDT identified but not understood.  Pass `tzinfos` argument in order to correctly return a timezone-aware datetime.  In a future version, this will raise an exception.
  warnings.warn("tzname {tzname} identified but not understood.  "


In [9]:
# Perform EDA
house.isna().sum()

street       0
city         0
zip          0
state        0
beds         0
baths        0
sq__ft       0
type         0
sale_date    0
price        0
latitude     0
longitude    0
dtype: int64

In [10]:
house.duplicated().sum()

3

In [18]:
#to get duplicated row info 
d_house = house.duplicated()
duplicates = house[d_house]
print(duplicates)

                street            city    zip state  beds  baths  sq__ft  \
342      4734 14TH AVE      SACRAMENTO  95820    CA     2      1     834   
404     8306 CURLEW CT  CITRUS HEIGHTS  95621    CA     4      2    1280   
601  7 CRYSTALWOOD CIR         LINCOLN  95648    CA     0      0       0   

            type                     sale_date   price   latitude   longitude  
342  Residential  Mon May 19 00:00:00 EDT 2008   68000  38.539447 -121.450858  
404  Residential  Mon May 19 00:00:00 EDT 2008  167293  38.715781 -121.298519  
601  Residential  Mon May 19 00:00:00 EDT 2008    4897  38.885962 -121.289436  


In [19]:
house.drop_duplicates()
print(house.iloc[[342,404]])

             street            city    zip state  beds  baths  sq__ft  \
342   4734 14TH AVE      SACRAMENTO  95820    CA     2      1     834   
404  8306 CURLEW CT  CITRUS HEIGHTS  95621    CA     4      2    1280   

            type                     sale_date   price   latitude   longitude  
342  Residential  Mon May 19 00:00:00 EDT 2008   68000  38.539447 -121.450858  
404  Residential  Mon May 19 00:00:00 EDT 2008  167293  38.715781 -121.298519  


In [20]:
#bed & sq_ft = 0 
bed_0 = house["beds"] == 0
print(house.loc[bed_0])


                       street             city    zip state  beds  baths  \
73              17 SERASPI CT       SACRAMENTO  95834    CA     0      0   
89           2866 KARITSA AVE       SACRAMENTO  95833    CA     0      0   
100     12209 CONSERVANCY WAY   RANCHO CORDOVA  95742    CA     0      0   
121       5337 DUSTY ROSE WAY   RANCHO CORDOVA  95742    CA     0      0   
126       2115 SMOKESTACK WAY       SACRAMENTO  95833    CA     0      0   
..                        ...              ...    ...   ...   ...    ...   
901  1530 TOPANGA LN Unit 204          LINCOLN  95648    CA     0      0   
917            501 POPLAR AVE  WEST SACRAMENTO  95691    CA     0      0   
934  1550 TOPANGA LN Unit 207          LINCOLN  95648    CA     0      0   
947     1525 PENNSYLVANIA AVE  WEST SACRAMENTO  95691    CA     0      0   
970             3557 SODA WAY       SACRAMENTO  95834    CA     0      0   

     sq__ft         type                     sale_date   price   latitude  \
73        

In [24]:
house['beds'].mean(axis = 0)

2.9116751269035532

In [ ]:
#Add the mean value into bed_0
#bed_0.assign

In [21]:
sq_ft_0 = house["sq__ft"] == 0
print(house.loc[sq_ft_0])

                       street             city    zip state  beds  baths  \
73              17 SERASPI CT       SACRAMENTO  95834    CA     0      0   
89           2866 KARITSA AVE       SACRAMENTO  95833    CA     0      0   
100     12209 CONSERVANCY WAY   RANCHO CORDOVA  95742    CA     0      0   
121       5337 DUSTY ROSE WAY   RANCHO CORDOVA  95742    CA     0      0   
126       2115 SMOKESTACK WAY       SACRAMENTO  95833    CA     0      0   
..                        ...              ...    ...   ...   ...    ...   
934  1550 TOPANGA LN Unit 207          LINCOLN  95648    CA     0      0   
947     1525 PENNSYLVANIA AVE  WEST SACRAMENTO  95691    CA     0      0   
966          9197 CORTINA CIR        ROSEVILLE  95678    CA     3      2   
970             3557 SODA WAY       SACRAMENTO  95834    CA     0      0   
973      2181 WINTERHAVEN CIR     CAMERON PARK  95682    CA     3      2   

     sq__ft         type                     sale_date   price   latitude  \
73        

In [ ]:
#check duplicated street name 
#house.duplicated('street').sum()

In [ ]:
#d_street = house.duplicated('street')
#duplicates_s = house[d_street]
#print(duplicates_s)

In [ ]:
#lamberton_1223 = house["street"] == "1223 LAMBERTON CIR"
#print(house.loc[lamberton_1223])

# Predict Price

We are going to predict `price` from available information.

## What is the Target Variable? Why?

In [ ]:
# ANSWER Price


## List all possible variables which might be Predictor variable.

In [ ]:
# ANSWER bed and sq_ft 

## Find correlation between variables

Find correlation between variables. Which is the best predictor? Why? State your reason.

In [ ]:
# ANSWER 

## What is the simplest way to predict price

So we have information for 985 houses. What is the easiest way to predict price of a house?

In [ ]:
# ANSWER

## Find Residuals

Find difference between actual value and predicted value. Use absolute value.

In [ ]:
# ANSWER

## Predict Using Linear Regression

[Simple Linear Regression](https://en.wikipedia.org/wiki/Simple_linear_regression#Fitting_the_regression_line)

$$\hat{y} = \beta_0 + \beta_1 x$$

where

> $$ \beta_1 = \frac{\sum_{i=1}^n (y_i - \bar{y} ) (x_i - \bar{x} )}{\sum_{i=1}^n (x_i - \bar{x})^2} $$

and

> $$ \beta_0 = \bar{y} - \beta_1\bar{x} $$

Given variables:
- $\bar{y}$ : the sample mean of observed values $Y$
- $\bar{x}$ : the sample mean of observed values $X$
- $s_Y$ : the sample standard deviation of observed values $Y$
- $s_X$ : the sample standard deviation of observed values $X$
- $r_{XY}$ : the sample Pearson correlation coefficient between observed $X$ and $Y$

### Define X, Y

- X = Predictor Variable 
- Y = Target Variable

In [30]:
# ANSWER
X = house['sq__ft']
Y = house['price']



### Calculate beta_1

$$ \beta_1 = \frac{\sum_{i=1}^n (y_i - \bar{y} ) (x_i - \bar{x} )}{\sum_{i=1}^n (x_i - \bar{x})^2} $$

In [ ]:
Y_mean = Y.mean()
X_mean = X.mean()

In [35]:
def Cal_beta_1():
    return np.sum(((Y - Y_mean) * (X - X_mean)) / ((X - X_mean) ** 2))


In [38]:
# ANSWER

    
    
beta_1 = (((Y - Y_mean) * (X - X_mean)).sum() / ((X - X_mean) ** 2)).sum() 
print(beta_1)

172766028.2342273


In [ ]:
175396.78421444073

### Calculate beta_0

$$ \beta_0 = \bar{y} - \beta_1\bar{x} $$

In [36]:
# ANSWER
def Cal_beta_0():
    return np.sum(Y_mean - (Cal_beta_1 * X_mean))

In [ ]:
beta_0 = ( Y_mean - (Cal_beta_1 * X_mean)

### Find Yhat (Predicted Value Using Beta 0 and Beta 1)

$$\hat{y} = \beta_0 + \beta_1 x$$

In [ ]:
# ANSWER


### Find Residuals

$$ \text{residual}_i = y_i - \hat{y}_i$$

In [ ]:
# ANSWER

### Find R-Square

The coefficient $R^2$ is defined as $$ R^2 = 1 - \frac{SS_{res}}{SS_{tot}} $$, where $$ SS_{res} $$ is the residual sum of squares $$ SS_{res} = \sum_{i=1}^n \left(y_i - \hat{y}_i\right)^2 $$ and $$ SS_{res} $$ is the total sum of squares $$ SS_{tot} = \sum_{i=1}^n \left(y_i - \bar{y}\right)^2 $$

The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse).

**Write a function to calculate R- Square value.**

The total sum of squares:

$$ SS_{tot} = \sum_{i=1}^n \left(y_i - \bar{y}\right)^2 $$

The residual sum of squares:

$$ SS_{res} = \sum_{i=1}^n \left(y_i - \hat{y}_i\right)^2 $$

$R^2$:

$$ R^2 = 1 - \frac{SS_{res}}{SS_{tot}} $$

In [ ]:
# ANSWER

### [BONUS] Draw Plot of Predicted Line

In [ ]:
# ANSWER

### [BONUS] Draw Plot of Predicted Line With Residuals

In [ ]:
# ANSWER

### Predict House Price

Let's assume we have information of following house:

- street:	1140 EDMONTON DR
- city:	SACRAMENTO
- zip:	95833
- state:	CA
- beds:	3
- baths:	2
- sq__ft:	1204
- type:	Residential

**Predict the price of this house.**

In [ ]:
# ANSWER

## Find Coefficient and Intercept using Linear SKLearn Regression

Using LinearRegression of sklearn.linear_model package find Coefficient and Intercept.

In [ ]:
# ANSWER

### Find R Score

In [ ]:
# ANSWER

# Splitting Data

In [ ]:
# ANSWER

## Create training and testing subsets

Hint: Using `train_test_split` Library

In [ ]:
# ANSWER

## Check Shape, Sample of Test Train Data

In [ ]:
# ANSWER

## Using Linear Regression Find The Score

1. Fit model using X_train, y_train
2. Find score using x_test, y_test

In [ ]:
# ANSWER

## Predict House Price

Let's assume we have information of following house:

- street:	1140 EDMONTON DR
- city:	SACRAMENTO
- zip:	95833
- state:	CA
- beds:	3
- baths:	2
- sq__ft:	1204
- type:	Residential

What is the predicted value of that house?

In [ ]:
# ANSWER

## Find the error

In [ ]:
# ANSWER